### ISSUE TITLE
Backreferences are sometimes parsed as octal escapes

### ISSUE THREAD

**Issue by mathiasbynens** [jviereck/regjsparser#70]
Originally reported by @nhahtdh, here: https://github.com/mathiasbynens/regexpu/issues/19

```bash
$ regjsparser '(\1)+\1\1'
{
  "type": "alternative",
  "body": [
    {
      "type": "quantifier",
      "min": 1,
      "max": null,
      "greedy": true,
      "body": [
        {
          "type": "group",
          "behavior": "normal",
          "body": [
            {
              "type": "value",
              "kind": "octal", // ←
              "codePoint": 1,
              "range": [
                1,
                3
              ],
              "raw": "\\1"
            }
          ],
          "range": [
            0,
            4
          ],
          "raw": "(\\1)"
        }
      ],
      "range": [
        0,
        5
      ],
      "raw": "(\\1)+"
    },
    {
      "type": "reference",
      "matchIndex": 1,
      "range": [
        5,
        7
      ],
      "raw": "\\1"
    },
    {
      "type": "reference",
      "matchIndex": 1,
      "range": [
        7,
        9
      ],
      "raw": "\\1"
    }
  ],
  "range": [
    0,
    9
  ],
  "raw": "(\\1)+\\1\\1"
}
```

**Comment by jviereck** [comment]
Copy the comment from mathiasbynens/regexpu#19:

---

Thanks for reporting this! From out of my head I think the current implementation of the `regjsparser` checks the so far parsed count of groups to decide if `\1` is a back-reference or a octal match.

Here are a few results from playing around with capturing matches from inside of Firefox Console:

```js
// Simple reduction of the reported bug.
r = /(\1)a/.exec('\x01a'); // -> [ "a", "" ]
r[1].length == 0 // true

// A little bit more advanced.
r = /(\2)(b)a/.exec('\x02ba');
r.length == 0;

// Q: What happens if the back-reference is larger than total number of left capturing parentheses?
// A: Interpret the \2 as an octal match.
/(\2)a/.exec('\x01a'); // -> null
/(\2)a/.exec('\x02a'); // ['\x02a', '\x02']
```

**Comment by jviereck** [comment]
Out of my head I can think about two ways to fix this bug:

1. Perform a first parse iteration to count the number of total left capturing parentheses in the RegExp and then on the second detailed parse use this total number to decide if `\1` is `\x01` or a back reference.
2. Perform the parsing as we do it at the moment, collect the positions a back-reference is encountered. After parsing the RegExp is finished, iterate over the collected back-references and decide how to handle them.

The second option seems easier to be implemented to me. Any objections or ideas?

**Comment by d10** [comment]
Yep, number 2 seems easier.

**Comment by mathiasbynens** [comment]
:+1: option 2!

**Comment by nhahtdh** [comment]
I think option 2 is good.

There is one thing I notice when reading the specs. ECMAScript 5 doesn't allow octal escape sequence in RegExp as far as I can see. However, your library seems to allow them. Which implementation of ECMAScript are you basing your grammar on then?

**Comment by mathiasbynens** [comment]
@nhahtdh Existing implementations support it, so we implement it. E.g. `/\1/.test('\x01') == true`. See https://javascript.spec.whatwg.org/#octal-escapes-in-regular-expression-literals.

**Comment by nhahtdh** [comment]
@mathiasbynens Lol... Seems that you just file that bug.

Anyway, I think you guys might want to set a baseline of what to implement. For example, currently, it seems to be ECMAScript + octal escape extension.

If you want to support existing implementation, you might want to consider `/{}/` or `/{}{}]/` or `/{abc}/`. From my testing, they seem to be supported on latest Chrome, Firefox and IE11, but they are invalid regex according to ECMAScript specs as far as I can see.

**Comment by mathiasbynens** [comment]
@nhahtdh Interesting – would you mind [filing a spec bug](https://bugs.ecmascript.org/) for those?

**Comment by nhahtdh** [comment]
@mathiasbynens It's not a spec bug IMO, just implementations' extension to the functionality of the regex defined in the standard.

(Basically, the implementations in the browsers allow:
- Unmatched closing `]` treated as literal
- Failed number quantifier treated as literal.)

It requires potentially long look-ahead for the case of `/{12,34xyz}/` (failed numbered quantifier treated as literal string), so I think it is not a good idea to include it in the specs.

**Comment by mathiasbynens** [comment]
@nhahtdh The spec not matching reality is usually a bug somewhere. It doesn’t hurt to file an issue for it. Pretty please? :)

Re: octal escapes in regular expressions, I forgot https://people.mozilla.org/~jorendorff/es6-draft.html#sec-regular-expressions-patterns already defines it (see `LegacyOctalEscapeSequence`).

**Comment by nhahtdh** [comment]
@mathiasbynens In the same section in your link (extension to RegExp), apart from octal escape extension, it also defines extension for failed number quantifier `{}` to be treated as literal character. (Check out the new definition for PatternCharacter and PatternCharacterNoBrace). Free `]` are not allowed even by the extension, though.

However, there seems to be a bug in Term_U (check [5.1.5](https://people.mozilla.org/%7Ejorendorff/es6-draft.html#sec-grammar-notation) for grammar notation)):
```
Term_U::
  Assertion_U
  Atom_U
  Atom_U Quantifier

Atom_U::
  PatternCharacter

PatternCharacter ::
  SourceCharacter but not one of
  ^ $ \ . * + ? ( ) [ ] |

```
Given `/a{3}/`, it seems that it will parse it as literal string `a { 3 }`, since the rules are ambiguous and the ambiguity is resolved by parsing the rules from top to bottom.

The intention of the author is to not change the behavior of U pattern, and only change that of the BMP pattern, but the author seems to mess up here.

**Comment by jviereck** [comment]
> 2. Perform the parsing as we do it at the moment, collect the positions a back-reference is encountered. After parsing the RegExp is finished, iterate over the collected back-references and decide how to handle them.

Looking at the code I doubt this is as simply doable as it sounds at first :/ The main problem is, that an input like `/\91/` can end up in two entries in the resulting AST. Substituting one entry with two entries is non trivial, as for example in the case of `/\91+/` the AST is at the top level made up of a quantifier node, that contains the to be decided `\91` part. As the number of total capturing groups is smaller than 9, the `\9` is to be interpreted as a single character like `/91+/`, which also means the `9` must move out of the quantifier node. I am not saying this cannot be done, but I feel the resulting code will do a lot of complex AST massaging and internal rewriting of `regjsparser` that might not be worth it for this case. ;)

As I think inputs like `/(\1)a/` are rare and I am in favor of a simple solution, here is my new idea:

3) Do a first parse of the input as we do it right now and record the decisions `refIdx <= closedCaptureCounter`. After the parsing has finished, we can check if any of these checks yields the wrong result as we now know the overall total count of capturing groups. If a false decision was made, reparse the input string once more and use the total count of capturing groups we have recorded during the first parsing phase.

So to sum up, in the worst case scenario the input string is parsed twice, which is unfortunate, but on the other side, this feels still more pleasant compared to doing a lot of internal AST reorganizing as with idea 1) or 2).
